### DISCUSSION TITLE
Loading Emulator with state programmatically

### DISCUSSION THREAD

**Discussion by rickhousley** [NationalSecurityAgency/ghidra#5649] [Debugger]
I'm interested in programmatically loading the emulator UI with register states, but I'm not sure how to get a handle on UI emulator context. I can identify how to create a new emulation using EmulatorHelper, but that's not quite what I'm looking for.

**Comment by nsadeveloper789** [comment]
Sorry I didn't see this for a year, apparently. In case it's still relevant to you or anyone else reading:

For scripting the UI emulator, it's probably best to use the `FlatDebuggerAPI`. You can access it from a `GhidraScript` by adding `implements FlatDebuggerAPI` to the class declaration for the script. If you're scripting from Python, I don't know how best to access it....

To launch an emulation session from the script, use `emulateLaunch`. For initializing register state, the method you want is `writeRegister` (there are several overloaded variants), but you might also look into `setControlMode`before that. That lets you choose whether the changes are written to the trace database, or just ephemerally to the emulator. Once your state is initialized, ensure the ControlMode is RW_EMULATOR, and then the `step` and `resume` methods will all start the emulator.

**Comment by Decryptortuning** [comment]
@rickhousley Are you releasing PERCH, or doing something similar? I Have been using the PER files for several years manually with some python scripting. Your tool looks useful however.
